In [17]:
#Regresion Lineal
#Arboles de decision
#random forest
import pandas as pd
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, f1_score

from sklearn.model_selection import train_test_split
import joblib

In [18]:
train_df = pd.read_csv('../Data/train.csv')
test_df = pd.read_csv('../Data/test.csv')
validation_df = pd.read_csv('../Data/validation.csv')

# Separate target column
train_X = train_df.drop(columns=['Depression'])
train_Y = train_df['Depression']

test_X = test_df.drop(columns=['Depression'])
test_Y = test_df['Depression']

validation_X = validation_df.drop(columns=['Depression'])
validation_Y = validation_df['Depression']

Se va a realizar un experimento donde se varian hiperparametros y se busca la mejor solucion

In [32]:
# Valores del hiperparámetro k (número de vecinos)
k_values = [3, 5, 7, 9]
weights_options = ['uniform', 'distance']
metric_options = ['euclidean', 'manhattan']

results = []

Se entrena el modelo con todos los posibles valores de hiperparametros y se selecciona el que tenga mejores métricas


In [37]:
from sklearn.metrics import confusion_matrix, precision_score, recall_score

# Loop a través de las combinaciones de hiperparámetros
for k in k_values:
    for weights in weights_options:
        for metric in metric_options:
            # Inicializar el clasificador k-NN con los hiperparámetros actuales
            model = KNeighborsClassifier(n_neighbors=k, weights=weights, metric=metric)

            # Entrenar el modelo
            model.fit(train_X, train_Y)

            # Realizar predicciones en el conjunto de validación
            validation_pred = model.predict(validation_X)

            # Calcular la matriz de confusión
            cm = confusion_matrix(validation_Y, validation_pred)

            # Calcular las métricas
            accuracy = accuracy_score(validation_Y, validation_pred)
            precision = precision_score(validation_Y, validation_pred, average='weighted')
            recall = recall_score(validation_Y, validation_pred, average='weighted')
            f1 = f1_score(validation_Y, validation_pred, average='weighted')

            # Guardar los resultados
            results.append({
                'k': k,
                'weights': weights,
                'metric': metric,
                'Accuracy': accuracy,
                'Precision': precision,
                'Recall': recall,
                'F1 Score': f1,
                'Confusion Matrix': cm
            })


In [38]:
# Convert the results to a DataFrame
results_df = pd.DataFrame(results)
results_df

k   weights     metric  Accuracy  Precision    Recall  F1 Score  \
0   3   uniform  euclidean  0.716981   0.716241  0.716981  0.716582   
1   3   uniform  manhattan  0.779874   0.778464  0.779874  0.778883   
2   3  distance  euclidean  0.716981   0.716241  0.716981  0.716582   
3   3  distance  manhattan  0.779874   0.778464  0.779874  0.778883   
4   5   uniform  euclidean  0.742138   0.739032  0.742138  0.739089   
5   5   uniform  manhattan  0.779874   0.778464  0.779874  0.778883   
6   5  distance  euclidean  0.742138   0.739032  0.742138  0.739089   
7   5  distance  manhattan  0.779874   0.778464  0.779874  0.778883   
8   7   uniform  euclidean  0.779874   0.777844  0.779874  0.777271   
9   7   uniform  manhattan  0.773585   0.771805  0.773585  0.769430   
10  7  distance  euclidean  0.779874   0.777844  0.779874  0.777271   
11  7  distance  manhattan  0.773585   0.771805  0.773585  0.769430   
12  9   uniform  euclidean  0.761006   0.758553  0.761006  0.758653   
13  9   uniform  manhattan  0.773585   0.771424  0.773585  0.771355   
14  9  distance  euclidean  0.761006   0.758553  0.761006  0.758653   
15  9  distance  manhattan  0.773585   0.771424  0.773585  0.771355   
16  3   uniform  euclidean  0.716981   0.716241  0.716981  0.716582   
17  3   uniform  manhattan  0.779874   0.778464  0.779874  0.778883   
18  3  distance  euclidean  0.716981   0.716241  0.716981  0.716582   
19  3  distance  manhattan  0.779874   0.778464  0.779874  0.778883   
20  5   uniform  euclidean  0.742138   0.739032  0.742138  0.739089   
21  5   uniform  manhattan  0.779874   0.778464  0.779874  0.778883   
22  5  distance  euclidean  0.742138   0.739032  0.742138  0.739089   
23  5  distance  manhattan  0.779874   0.778464  0.779874  0.778883   
24  7   uniform  euclidean  0.779874   0.777844  0.779874  0.777271   
25  7   uniform  manhattan  0.773585   0.771805  0.773585  0.769430   
26  7  distance  euclidean  0.779874   0.777844  0.779874  0.777271   
27  7  distance  manhattan  0.773585   0.771805  0.773585  0.769430   
28  9   uniform  euclidean  0.761006   0.758553  0.761006  0.758653   
29  9   uniform  manhattan  0.773585   0.771424  0.773585  0.771355   
30  9  distance  euclidean  0.761006   0.758553  0.761006  0.758653   
31  9  distance  manhattan  0.773585   0.771424  0.773585  0.771355   
32  3   uniform  euclidean  0.716981   0.716241  0.716981  0.716582   
33  3   uniform  manhattan  0.779874   0.778464  0.779874  0.778883   
34  3  distance  euclidean  0.716981   0.716241  0.716981  0.716582   
35  3  distance  manhattan  0.779874   0.778464  0.779874  0.778883   
36  5   uniform  euclidean  0.742138   0.739032  0.742138  0.739089   
37  5   uniform  manhattan  0.779874   0.778464  0.779874  0.778883   
38  5  distance  euclidean  0.742138   0.739032  0.742138  0.739089   
39  5  distance  manhattan  0.779874   0.778464  0.779874  0.778883   
40  7   uniform  euclidean  0.779874   0.777844  0.779874  0.777271   
41  7   uniform  manhattan  0.773585   0.771805  0.773585  0.769430   
42  7  distance  euclidean  0.779874   0.777844  0.779874  0.777271   
43  7  distance  manhattan  0.773585   0.771805  0.773585  0.769430   
44  9   uniform  euclidean  0.761006   0.758553  0.761006  0.758653   
45  9   uniform  manhattan  0.773585   0.771424  0.773585  0.771355   
46  9  distance  euclidean  0.761006   0.758553  0.761006  0.758653   
47  9  distance  manhattan  0.773585   0.771424  0.773585  0.771355   

        Confusion Matrix  
0   [[40, 23], [22, 74]]  
1   [[44, 19], [16, 80]]  
2   [[40, 23], [22, 74]]  
3   [[44, 19], [16, 80]]  
4   [[39, 24], [17, 79]]  
5   [[44, 19], [16, 80]]  
6   [[39, 24], [17, 79]]  
7   [[44, 19], [16, 80]]  
8   [[42, 21], [14, 82]]  
9   [[40, 23], [13, 83]]  
10  [[42, 21], [14, 82]]  
11  [[40, 23], [13, 83]]  
12  [[41, 22], [16, 80]]  
13  [[42, 21], [15, 81]]  
14  [[41, 22], [16, 80]]  
15  [[42, 21], [15, 81]]  
16  [[40, 23], [22, 74]]  
17  [[44, 19], [16, 80]]  
18  [[40, 2

In [39]:
# Seleccionar los mejores parámetros con base en F1 Score (o Accuracy)
best_params = results_df.loc[results_df['F1 Score'].idxmax()]
best_params

k                                      3
weights                          uniform
metric                         manhattan
Accuracy                        0.779874
Precision                       0.778464
Recall                          0.779874
F1 Score                        0.778883
Confusion Matrix    [[44, 19], [16, 80]]
Name: 1, dtype: object

Se entrena el modelo con los mejores parámetros.

In [40]:
# Entrenar el mejor modelo con los hiperparámetros del mejor resultado
best_model = KNeighborsClassifier(
    n_neighbors=int(best_params['k']),
    weights=best_params['weights'],
    metric=best_params['metric']
)

best_model.fit(train_X, train_Y)

KNeighborsClassifier(metric='manhattan', n_neighbors=3)

## Finalmente se usan los datos de "test" para validar las métricas del modelo.

In [41]:
# Evaluate the model on the test set
test_pred = best_model.predict(test_X)
test_mse = mean_squared_error(test_Y, test_pred)
test_r2 = r2_score(test_Y, test_pred)

# Display the final model metrics on the test set
print(f"Test MSE: {test_mse}")
print(f"Test R2: {test_r2}")

Test MSE: 0.2562814070351759
Test R2: -0.08568677792041113


Como ejemplo: se tiene un nuevo inmueble dado por los siguientes datos y se debe predecir el precio.

In [42]:
new_data = {
  "Age": [26.0],
  "Academic Pressure": [4.0],
  "CGPA": [8.78],
  "Study Satisfaction": [2.0],
  "Work/Study Hours": [5.0],
  "Financial Stress": [2.0],
  "Gender_Male": [0.0],
  "Dietary Habits_Moderate": [1.0],
  "Dietary Habits_Unhealthy": [0.0],
  "Degree_B.Arch": [0.0],
  "Degree_B.Com": [1.0],
  "Degree_B.Ed": [0.0],
  "Degree_B.Pharm": [0.0],
  "Degree_B.Tech": [0.0],
  "Degree_BA": [0.0],
  "Degree_BBA": [0.0],
  "Degree_BCA": [0.0],
  "Degree_BE": [0.0],
  "Degree_BHM": [0.0],
  "Degree_BSc": [0.0],
  "Degree_LLB": [0.0],
  "Degree_LLM": [0.0],
  "Degree_M.Com": [0.0],
  "Degree_M.Ed": [0.0],
  "Degree_M.Pharm": [0.0],
  "Degree_M.Tech": [0.0],
  "Degree_MA": [0.0],
  "Degree_MBA": [0.0],
  "Degree_MBBS": [0.0],
  "Degree_MCA": [0.0],
  "Degree_MD": [0.0],
  "Degree_ME": [0.0],
  "Degree_MSc": [0.0],
  "Degree_PhD": [0.0],
  "Have you ever had suicidal thoughts ?_Yes": [1.0],
  "Family History of Mental Illness_Yes": [0.0],
  "City_Ahmedabad": [0.0],
  "City_Bangalore": [0.0],
  "City_Bhopal": [0.0],
  "City_Chennai": [0.0],
  "City_Delhi": [0.0],
  "City_Faridabad": [0.0],
  "City_Ghaziabad": [0.0],
  "City_Hyderabad": [0.0],
  "City_Indore": [0.0],
  "City_Jaipur": [0.0],
  "City_Kalyan": [0.0],
  "City_Kanpur": [0.0],
  "City_Kolkata": [0.0],
  "City_Lucknow": [0.0],
  "City_Ludhiana": [0.0],
  "City_Meerut": [0.0],
  "City_Mumbai": [0.0],
  "City_Nagpur": [0.0],
  "City_Nashik": [0.0],
  "City_Patna": [0.0],
  "City_Pune": [0.0],
  "City_Rajkot": [0.0],
  "City_Srinagar": [0.0],
  "City_Surat": [0.0],
  "City_Thane": [0.0],
  "City_Vadodara": [0.0],
  "City_Varanasi": [0.0],
  "City_Vasai-Virar": [0.0],
  "City_Visakhapatnam": [0.0],
  "Sleep Duration_'7-8 hours'": [1.0],
  "Sleep Duration_'Less than 5 hours'": [0.0],
  "Sleep Duration_'More than 8 hours'": [0.0]
}



# Convertir el nuevo dato en un DataFrame
new_df = pd.DataFrame(new_data)

# Usar el modelo entrenado (best_model) para hacer la predicción
predicted_price = best_model.predict(new_df)

# Mostrar el resultado de la predicción
print(f"Predicted: {predicted_price[0]}")

Predicted: 1


In [43]:
new_data = {
  "Age": [33.0],
  "Academic Pressure": [5.0],
  "CGPA": [9.54],
  "Study Satisfaction": [1.0],
  "Work/Study Hours": [7.0],
  "Financial Stress": [2.0],
  "Gender_Male": [1.0],
  "Dietary Habits_Moderate": [0.0],
  "Dietary Habits_Unhealthy": [0.0],
  "Degree_B.Arch": [0.0],
  "Degree_B.Com": [0.0],
  "Degree_B.Ed": [0.0],
  "Degree_B.Pharm": [1.0],
  "Degree_B.Tech": [0.0],
  "Degree_BA": [0.0],
  "Degree_BBA": [0.0],
  "Degree_BCA": [0.0],
  "Degree_BE": [0.0],
  "Degree_BHM": [0.0],
  "Degree_BSc": [0.0],
  "Degree_LLB": [0.0],
  "Degree_LLM": [0.0],
  "Degree_M.Com": [0.0],
  "Degree_M.Ed": [0.0],
  "Degree_M.Pharm": [0.0],
  "Degree_M.Tech": [0.0],
  "Degree_MA": [0.0],
  "Degree_MBA": [0.0],
  "Degree_MBBS": [0.0],
  "Degree_MCA": [0.0],
  "Degree_MD": [0.0],
  "Degree_ME": [0.0],
  "Degree_MSc": [0.0],
  "Degree_PhD": [0.0],
  "Have you ever had suicidal thoughts ?_Yes": [1.0],
  "Family History of Mental Illness_Yes": [0.0],
  "City_Ahmedabad": [0.0],
  "City_Bangalore": [0.0],
  "City_Bhopal": [0.0],
  "City_Chennai": [0.0],
  "City_Delhi": [0.0],
  "City_Faridabad": [0.0],
  "City_Ghaziabad": [0.0],
  "City_Hyderabad": [0.0],
  "City_Indore": [0.0],
  "City_Jaipur": [0.0],
  "City_Kalyan": [0.0],
  "City_Kanpur": [0.0],
  "City_Kolkata": [0.0],
  "City_Lucknow": [0.0],
  "City_Ludhiana": [0.0],
  "City_Meerut": [0.0],
  "City_Mumbai": [0.0],
  "City_Nagpur": [0.0],
  "City_Nashik": [0.0],
  "City_Patna": [0.0],
  "City_Pune": [0.0],
  "City_Rajkot": [0.0],
  "City_Srinagar": [0.0],
  "City_Surat": [0.0],
  "City_Thane": [0.0],
  "City_Vadodara": [0.0],
  "City_Varanasi": [0.0],
  "City_Vasai-Virar": [0.0],
  "City_Visakhapatnam": [0.0],
  "Sleep Duration_'7-8 hours'": [0.0],
  "Sleep Duration_'Less than 5 hours'": [0.0],
  "Sleep Duration_'More than 8 hours'": [0.0]
}




# Convertir el nuevo dato en un DataFrame
new_df = pd.DataFrame(new_data)

# Usar el modelo entrenado (best_model) para hacer la predicción
predicted_price = best_model.predict(new_df)

# Mostrar el resultado de la predicción
print(f"Predicted: {predicted_price[0]}")

Predicted: 1


En nuestro modelo de clasificación utilizando k-Nearest Neighbors (k-NN), obtuvimos una matriz de confusión con 44 verdaderos negativos y 80 verdaderos positivos, junto a 19 falsos positivos y 16 falsos negativos. A partir de estos resultados, calculamos una precisión del 77.85%, lo que indica que el 77.85% de nuestras predicciones positivas fueron correctas. La exactitud general del modelo fue del 77.99%, lo que significa que casi el 78% de todas las predicciones fueron acertadas. Además, el recall también fue del 77.99%, mostrando que el modelo identificó correctamente la mayoría de las instancias positivas. Finalmente, el F1 Score fue de 77.89%, evidenciando un buen equilibrio entre precisión y recall. Estos resultados nos permiten evaluar la solidez del modelo, aunque también resaltan la necesidad de seguir afinando los parámetros para reducir los errores en futuras predicciones.